### DISCUSSION TITLE
Communication with slurm heterogeneous jobs

### DISCUSSION THREAD

**Discussion by echaussidon** [mpi4py/mpi4py#200] [Q&A] [answered]
Hi, 

I have an issue when I use mpi4py with slurm heterogeneous jobs.  I can communicate between ranks in each component itself but I'm not able to communicate with all the rank together. (All the ranks seem to be in the same MPI_COMM_WORLD) There is no error but nothing happens. (Same problem with comm.Barrier() or comm.allgather() ect ...)

What I try to do : Creating with slurm heterogeneous jobs (https://slurm.schedmd.com/heterogeneous_jobs.html#job_steps) two components. the first component is the rank 0, the second component is rank 1 and 2. In the following exemple MPI.COMM_WORLD.size = 3 as expected. Then I try to communicate between rank 1 and rank 2. No problem. And then I try to send something to the rank 0 and nothing happens..  

Maybe the problem comes from port detection or someting like that but I was not able to find any information on the web.. Do you know what can I do to test it in more detail ?  or maybe someone already use to this problem ?

I'm working in NERSC with mpich 3.3 and Here the code that I use, test.py file : 
```
from mpi4py import MPI
import numpy as np

comm = MPI.COMM_WORLD
rank = comm.Get_rank()

print(f"Hello from {rank} / {comm.size}", flush=True)

tt = np.zeros(3, dtype='i8')
to_send = np.array([1, 2, 3], dtype='i8')

# This work
if rank == 1:  comm.Send(to_send, dest=2, tag=1)
if rank == 2: comm.Recv(tt, source=1, tag=1)

print(rank, tt, flush=True)

# 
if rank == 2:  comm.Send(to_send, dest=0, tag=2)
if rank == 0:  comm.Recv(tt, source=2, tag=2)

print(rank, tt, flush=True)
```

and the script to launch the job:

```
#!/bin/bash

#SBATCH --nodes=1 --ntasks=1 --cpus-per-task=68
#SBATCH --time=00:01:00 --qos=debug --constraint=knl --account=desi
#SBATCH --job-name=test --output="%x-%j.out"

#SBATCH hetjob
#SBATCH --nodes=1 --ntasks=2 --cpus-per-task=1
#SBATCH --time=00:01:00 --qos=debug --constraint=knl --account=desi

module load python
srun  --het-group=0,1 python test.py
```

**Comment by dalcinl** [comment]
I cannot tell what's going on in your side, but this is most likely not an mpi4py issue. I would recommend to post your question to mpi4py's mailing list at Google Groups or other online forums. Also, why don't you just ask to NERSC support?

**Comment by echaussidon** [comment, marked as answer]
Hi, 

Sorry I was exhausted, after long research... 

For people, who maybe will get the same issue, in NERSC it does not work with `srun --het-group=0,1 python test.py` command but this work with `srun python test.py : python test.py`

I'm sorry to raise this issue. 

Regards, 

Edmond
